mozilla_version/mobile.py

```diff
@@ -42,30 +46,97 @@ class MobileVersion(BaseVersion):
+        'patch_number', 'beta_number', 'release_candidate_number', 'build_number'
     )
 
-    _ALL_NUMBERS = _MANDATORY_NUMBERS + _OPTIONAL_NUMBERS
+    _ALL_NUMBERS = BaseVersion._MANDATORY_NUMBERS + _OPTIONAL_NUMBERS
+
+    _FIRST_VERSION_TO_FOLLOW_GECKO_PATTERN = 104
 
+    build_number = attr.ib(type=int, converter=strictly_positive_int_or_none, default=None)
     beta_number = attr.ib(type=int, converter=strictly_positive_int_or_none, default=None)
+    is_nightly = attr.ib(type=bool, default=False)
     release_candidate_number = attr.ib(
         type=int, converter=strictly_positive_int_or_none, default=None
     )
     version_type = attr.ib(init=False, default=attr.Factory(_find_type, takes_self=True))
 
+    def __attrs_post_init__(self):
+        """Ensure attributes are sane all together."""
+        error_messages = []
+
+        if self.is_gecko_pattern:
+            error_messages.extend([
+                pattern_message
+                for condition, pattern_message in ((
+                    self.beta_number is not None and self.patch_number is not None,
+                    'Beta number and patch number cannot be both defined',
+                ), (
+                    self.release_candidate_number is not None,
+                    'Release candidate number cannot be defined starting Mobile v104',
+                ), (
+                    self.minor_number == 0 and self.patch_number == 0,
+                    'Minor number and patch number cannot be both equal to 0',
+                ), (
+                    self.minor_number != 0 and self.patch_number is None,
+                    'Patch number cannot be undefined if minor number is greater than 0',
+                ))
+                if condition
+            ])
+        else:
+            error_messages.extend([
+                pattern_message
+                for condition, pattern_message in ((
+                    self.patch_number is None,
+                    'Patch number must be defined before Mobile v104',
+                ), (
+                    self.is_nightly,
+                    'Nightlies are not supported until Mobile v104',
+                ))
+                if condition
+            ])
+
+        if error_messages:
+            raise PatternNotMatchedError(self, patterns=error_messages)
+
+    @classmethod
+    def parse(cls, version_string):
+        """Construct an object representing a valid Firefox version number."""
+        mobile_version = super(MobileVersion, cls).parse(
+            version_string, regex_groups=('is_nightly',)
+        )
+
+        # Betas are supported in both the old and the gecko pattern. Let's make sure
+        # the string we got follows the right rules
+        if mobile_version.is_beta:
+            if mobile_version.is_gecko_pattern and '-beta.' in version_string:
+                raise PatternNotMatchedError(
+                    mobile_version, ['"-beta." cannot be used starting Mobile v104']
```

Comment:
nit: awkward message

Reply:
Changed into `"-beta." can only be used before Mobile v104`. Let me know if you'd like me to rephrase it. I'm happy to do it in a follow-up.